cellmlmanip/model.py

```diff
@@ -619,6 +619,205 @@ def variables(self):
         """ Returns an iterator over this model's variable symbols. """
         return self._name_to_symbol.values()
 
+    def add_input(self, name, units, is_output=False):
```

Comment:
Um, I didn't claim that! Or at least I didn't mean to :)

Personally I like the `add_input` and `add_output` names, as they match what I used before and how I think about this. I'm open to others (@MauriceHendrix ?) suggesting something better, but haven't seen a concrete suggestion yet.

From an FC point of view, specifying the variable either by passing a `VariableDummy` or ontology term would be most useful, so those would be the two options I'd suggest for a first argument called `variable`. Giving the units as an object as done here is also what we'd want.

It's possible that Chaste code-gen would need to convert a variable that doesn't have an annotation, so the option of both might help @MauriceHendrix . And even if all variables do have annotations, the caller might well already know the `VariableDummy` so it makes sense to avoid another lookup.

Reply:
what about 'add_and_convert_input/output'

Reply:
Hmm, that gives me some ideas. We could do `convert_variable` (or even `maybe_convert_variable`!), and either have two methods (`convert_variable_as_input`) or have the direction as an argument (`maybe_convert_variable(var, units, direction=DataFlowDirection.TO_MODEL)`) with a `DataFlowDirection` enum having options `TO_MODEL` and `FROM_MODEL`. Does that make the API clearer?

It might also be handy to make the return value either the original or newly added variable, depending on whether a conversion was needed?

Reply:
I like the idea of some way of telling the user 'nothing changed'

Reply:
@MauriceHendrix @MichaelClerx any preferences on naming?

I propose `maybe_convert_variable(var, units, direction)` with 3 required positional arguments.

Reply:
I'm happy to let @skeating and @jonc125 decide the names!

Reply:
So conclusion:

function is 
var = convert_variable(DummyVariable/ontology term/string name, Unit object, direction = {IN, OUT} enum)

returns the converted var - which may be same as original argument if no conversion was necessary
throws exceptions for
- Variable not in model
- Unit conversion impossible (ie m -> kg)

enum will be nicer - above just for clarity

we dont allow a string name for the unit

Everyone happy ??

Reply:
Yes to all of this, except that I'd suggest not allowing the 'string name' option for variable - too likely to cause confusion between whether a string is a URI (ontology term) or name.